Allow extra arguments for vos release

marciobarbosa opened the pull request:
The commands in the vos command suite are the administrative interface
to the Volume Server and Volume Location Server. As we know, CellCC
takes advantage of these commands to synchronize volumes between AFS
cells. Unfortunately, the system administrator does not have the freedom
to specify which options will be passed to vos.

To mitigate the limitation mentioned above, accept extra arguments for
vos release. To make it possible, read the cellcc.json configuration
file to check if any extra flag was provided by the user.

marciobarbosa commented:
Pushing the first version to get some feedback.

Thanks!

marciobarbosa commented on lib/AFS/CellCC/Restore.pm:
```diff
@@ -471,10 +471,18 @@ _do_release($$) {
+        $args{cell} = $job->{dst_cell};
+
+        while (my ($key, $value) = each %flags) {
+            $args{$key} = $value;
```
Maybe, I should call supportsArgument (from afs-command) here.

adeason replied:
I don't think that's needed, unless the failure mode is really bad/confusing when AFS::Command gets an unrecognized argument. What happens now?

marciobarbosa replied:
The failure mode is not confusing (we get a message like: "Unsupported arguments"). So, yes, I think you are correct.

adeason commented on doc/pod1/cellcc_config.pod:
```diff
@@ -196,6 +196,12 @@ authenticate to AFS in order to run authenticated B<vos> commands.
+
+This can be used to pass extra options to vos release. As an example, if the
+system administrator wants to enable dpf, we would have to set
+B<vos/flags/release/dpf>.
```
Don't use dpf as an example; that's an experimental feature that only exists in site-specific deployments. We should also provide an actual example of what this looks like in the config file; it's not completely clear how to give options as key/val pairs (I know the existing documentation isn't good about providing concrete examples; we should start doing better). Suggested text:

This can be used to pass extra options to the "vos release" command line when CellCC releases a volume. The options must be given as an associative array as accepted by AFS::Command, with the option name as the key and its argument as the value. For example, to pass "-encrypt -config /usr/vice/etc", specify the following:

{
  vos: {
    flags: {
      release: {
        encrypt: True,
        config: "/usr/vice/etc",
      }
    }
  }
}

...or fix that with whatever syntax actually works in the config file :)

marciobarbosa replied:
Done.

adeason commented on lib/AFS/CellCC/Config.pm:
```diff
@@ -118,6 +118,8 @@ my @directives = (
     { key => 'vos/princ', default => undef, },
     { key => 'vos/keytab', default => "$CONF_DIR/vos.keytab", },
+    { key => 'vos/flags', type => 'HASH', default => {}, },
+    { key => qr:^vos/flags/[^/]+/*:, default => undef, },
```
this commit doesn't seem to use the plain "vos/flags" option, and we don't honor flags for arbitrary vos ops, just for release. I think this should just be:

{ key => 'vos/flags/release', type => 'HASH', default -> {}, } ,

Adding configurable flags to all individual vos operations and to vos generically makes sense if you want to do that (but it's not needed right now). But if we want to add generic vos flags to pass for all operations, you can't have "vos/flags" and "vos/flags/release", since "release" looks like a flag to add. So maybe this should be the other way around? That is, make "vos/release/flags", and eventually we could add a generic "vos/flags" later on. If that makes sense to you?

marciobarbosa replied:
Make sense (done).

adeason reviewed (commented):
Another way to implement this kind of thing would be to make vos_auth/vos_unauth return an object of our own class that wraps the AFS::Command object, and inserts our own flags when the caller calls e.g. $vos->release. That would be a much easier approach when adding this kind of functionality to the other vos sub-commands, but if we're only doing this for "vos release", that doesn't need to be done yet.

adeason commented on lib/AFS/CellCC/Config.pm:
```diff
@@ -107,6 +107,8 @@ my @directives = (
     { key => 'restore/queues', type => 'HASH', default => {}, },
     { key => qr:^restore/queues/[^/]+/max-parallel$:, default => 1, },
+    { key => qr:^restore/queues/[^/]+/release/flags$:, type => 'HASH', default => {}, },
+    { key => qr:^restore/queues/[^/]+/release/flags/+:, default => undef, },
```
is this line needed? I would think this would be handled by the previous line (or the previous line is unnecessary), but I may not be remembering the config parsing/checking right. Do we actually need both?

if we do need this, that regex I think should be qr:^restore/queues/[^/]+/release/flags/[^/]+$:

marciobarbosa replied:
I think we need both. Without line 111, we get messages like (from the check-server):

Configuration error: Unknown config directive 'restore/queues/<queue>/release/flags/<flag>'.

Line 110 seems to be unnecessary. However, if we do not have an entry for the default queue (restore/queues/default/release/flags), config_get (Restore.pm:475) will fail. With line 110, if we do not have an entry for the default queue, we get an empty hash (and config_get does not fail).

> if we do need this, that regex I think should be
> qr:^restore/queues/[^/]+/release/flags/[^/]+$:

Agreed.

marciobarbosa committed:
Allow extra arguments for vos release

The commands in the vos command suite are the administrative interface
to the Volume Server and Volume Location Server. As we know, CellCC
takes advantage of these commands to synchronize volumes between AFS
cells. Unfortunately, the system administrator currently cannot
specify which options will be passed to vos.

To improve this, accept extra vos arguments in cellcc.json for the
current queue. In the future we may do this for most/all vos
operations, but for now only do this for 'vos release'.

adeason committed:
Fix cellcc_config manpage formatting

These two lines were missing a '/' after the qname. Add the missing
slash.

adeason commented on lib/AFS/CellCC/Config.pm:
```diff
@@ -107,6 +107,8 @@ my @directives = (
     { key => 'restore/queues', type => 'HASH', default => {}, },
     { key => qr:^restore/queues/[^/]+/max-parallel$:, default => 1, },
+    { key => qr:^restore/queues/[^/]+/release/flags$:, type => 'HASH', default => {}, },
+    { key => qr:^restore/queues/[^/]+/release/flags/[^/]+$:, },
```
after some offline discussion, I agree with these lines (I was thinking maybe a comment should be added about this... but we already do this for other stuff; I was just being dumb about it). However, it's also possible I think that this could cause problems in the future if someone wants to specify an array or hash value for a vos option. e.g. if there is a CMD_LIST vos parameter that the administrator wants to give, we'll need to give an array in the json, but this will probably fail config checking, since we'll be expecting a SCALAR ref and we'll get an ARRAY ref in the config.

I think I'm just going to not care until that actually becomes a problem, but I just wanted to note it here that that exists.
